xdem/coreg/base.py

```diff
@@ -725,16 +722,56 @@ def is_affine(self) -> bool:
 
         return self._is_affine
 
+    def _get_subsample_on_valid_mask(self, valid_mask: NDArrayb, verbose: bool = False) -> NDArrayb:
+        """
+        Get mask of values to subsample on valid mask.
+
+        :param valid_mask: Mask of valid values (inlier and not nodata).
+        """
+
+        # This should never happen
+        if self._meta["subsample"] is None:
+            raise ValueError("Subsample should have been defined in metadata before reaching this class method.")
+
+        # If subsample is not equal to one, subsampling should be performed.
+        elif self._meta["subsample"] != 1.0:
+
+            # Build a low memory masked array with invalid values masked to pass to subsampling
+            ma_valid = np.ma.masked_array(data=np.ones(np.shape(valid_mask), dtype=bool), mask=~valid_mask)
+            # Take a subsample within the valid values
+            indices = gu.raster.subsample_array(
+                ma_valid,
+                subsample=self._meta["subsample"],
+                return_indices=True,
+                random_state=self._meta["random_state"],
+            )
+
+            # We return a boolean mask of the subsample within valid values
+            subsample_mask = np.zeros(np.shape(valid_mask), dtype=bool)
+            subsample_mask[indices[0], indices[1]] = True
+        else:
+            # If no subsample is taken, use all valid values
+            subsample_mask = valid_mask
+
+        if verbose:
+            print(
+                "Using a subsample of {} among {} valid values.".format(
+                    np.count_nonzero(valid_mask), np.count_nonzero(subsample_mask)
+                )
+            )
+
+        return subsample_mask
+
     def fit(
         self: CoregType,
         reference_dem: NDArrayf | MArrayf | RasterType,
         dem_to_be_aligned: NDArrayf | MArrayf | RasterType,
-        inlier_mask: NDArrayf | Mask | None = None,
+        inlier_mask: NDArrayb | Mask | None = None,
```

Comment:
Does this have a check if the user passes a uint8 mask? It could just be converted or error out.

Reply:
I think the check is still the same one you wrote originally, looks like we only accept `bool`:
```python
            assert inlier_mask.dtype == bool, f"Invalid mask dtype: '{inlier_mask.dtype}'. Expected 'bool'"
```

Reply:
This joins #293!
Added a comment: https://github.com/GlacioHack/xdem/issues/293#issuecomment-1734573084